flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java

```diff
@@ -388,6 +394,127 @@ public void vertexIsNotAffectedByOutdatedDeployment() {
+		scheduler.updateTaskExecutionState(new TaskExecutionState(jobGraph.getJobID(), attemptId, ExecutionState.RUNNING));
+
+		final CheckpointCoordinator checkpointCoordinator = getCheckpointCoordinator(scheduler);
+
+		// register a stateful master hook to help verify state restore
+		final TestMasterHook masterHook = new TestMasterHook("testHook");
+		checkpointCoordinator.addMasterHook(masterHook);
+
+		// complete one checkpoint for state restore
+		checkpointCoordinator.triggerCheckpoint(System.currentTimeMillis(),  false);
+		acknowledgeCheckpoint(checkpointCoordinator, jobGraph.getJobID(), attemptId);
+
+		scheduler.updateTaskExecutionState(new TaskExecutionState(jobGraph.getJobID(), attemptId, ExecutionState.FAILED));
+		taskRestartExecutor.triggerScheduledTasks();
+		assertThat(masterHook.getRestoreCount(), is(equalTo(1)));
+	}
+
+	@Test
+	public void failGlobalWhenRestoringStateFails() throws Exception {
+		final JobGraph jobGraph = singleNonParallelJobVertexJobGraph();
+		final JobVertex onlyJobVertex = getOnlyJobVertex(jobGraph);
+		enableCheckpointing(jobGraph);
+
+		final DefaultScheduler scheduler = createSchedulerAndStartScheduling(jobGraph);
+
+		final ArchivedExecutionVertex onlyExecutionVertex = Iterables.getOnlyElement(scheduler.requestJob().getAllExecutionVertices());
+		final ExecutionAttemptID attemptId = onlyExecutionVertex.getCurrentExecutionAttempt().getAttemptId();
+		scheduler.updateTaskExecutionState(new TaskExecutionState(jobGraph.getJobID(), attemptId, ExecutionState.RUNNING));
+
+		final CheckpointCoordinator checkpointCoordinator = getCheckpointCoordinator(scheduler);
+
+		// register a master hook to fail state restore
+		final TestMasterHook masterHook = new TestMasterHook("testHook");
+		masterHook.enableFailOnRestore();
+		checkpointCoordinator.addMasterHook(masterHook);
+
+		// complete one checkpoint for state restore
+		checkpointCoordinator.triggerCheckpoint(System.currentTimeMillis(),  false);
+		acknowledgeCheckpoint(checkpointCoordinator, jobGraph.getJobID(), attemptId);
+
+		scheduler.updateTaskExecutionState(new TaskExecutionState(jobGraph.getJobID(), attemptId, ExecutionState.FAILED));
+		taskRestartExecutor.triggerScheduledTasks();
+		final List<ExecutionVertexID> deployedExecutionVertices = testExecutionVertexOperations.getDeployedVertices();
+
+		// the first task failover should be skipped on state restore failure
+		final ExecutionVertexID executionVertexId = new ExecutionVertexID(onlyJobVertex.getID(), 0);
+		assertThat(deployedExecutionVertices, contains(executionVertexId));
+
+		// a global failure should be triggered on state restore failure
+		masterHook.disableFailOnRestore();
+		taskRestartExecutor.triggerScheduledTasks();
+		assertThat(deployedExecutionVertices, contains(executionVertexId, executionVertexId));
+	}
+
+	private void acknowledgeCheckpoint(
+			final CheckpointCoordinator checkpointCoordinator,
+			final JobID jobID,
+			final ExecutionAttemptID attemptId) throws Exception {
```

Comment:
How about changing this method to acknowledge all executions? I think it would better facilitate code reuse.

Reply:
Ok. done by changing it to `acknowledgePendingCheckpoint(scheduler, checkpointId)` to acknowledge all the executions for the given checkpoint.